pkg/controller/snapshotter/snapshotter.go

```diff
@@ -347,204 +345,310 @@ func (vs *volumeSnapshotter) deleteSnapshot(spec *v1.PersistentVolumeSpec, sourc
-				snapshot.Status.Conditions[index].Status == v1.ConditionTrue) ||
-				snapshot.Status.Conditions[index].Type == crdv1.VolumeSnapshotConditionError) {
-			glog.Infof("getSnapshotCreateFunc: Snapshot %s is Ready.", snapshotName)
-			return nil
-		} else if len(snapshot.Status.Conditions) > 0 &&
-			snapshot.Status.Conditions[index].Type == crdv1.VolumeSnapshotConditionPending &&
-			(snapshot.Status.Conditions[index].Status == v1.ConditionTrue ||
-				snapshot.Status.Conditions[index].Status == v1.ConditionUnknown) {
-			glog.Infof("getSnapshotCreateFunc: Snapshot %s is Pending.", snapshotName)
-			// Query the volume plugin for the status of the snapshot with snapshot id
-			// from VolumeSnapshotData object.
-			// Add snapshot to Actual State of World when snapshot is Ready.
-			err := vs.waitForPluginSnapshot(snapshotName, snapshot)
-			if err != nil {
-				return fmt.Errorf("Failed to create snapshot %s.", snapshotName)
-			}
-			glog.Infof("getSnapshotCreateFunc: Snapshot %s created successfully.", snapshotName)
-			return nil
-		} else {
-			glog.Infof("getSnapshotCreateFunc: Creating snapshot %s ...", snapshotName)
-			pv, err := vs.getPVFromVolumeSnapshot(snapshotName, snapshot)
-			if err != nil {
-				return err
-			}
-			pvName := pv.Name
-
-			// Retrieve metadata from VolumeSnapshot object if existing,
-			// otherwise create metadata and update VolumeSnapshot object.
-
-			// Look for VolumeSnapshot metadata ...
-			var tags *map[string]string
-			var bFoundVolumeSnapshotMetadata bool = false
-			if snapshot.Metadata.Name != "" && snapshot.Metadata.Namespace != "" {
-				if snapshot.Metadata.Labels != nil {
-					_, ok := snapshot.Metadata.Labels["Timestamp"]
-					if ok {
-						glog.Infof("Found Metadata Labels in VolumeSnapshot object: [%#v].", snapshot.Metadata.Labels)
-						bFoundVolumeSnapshotMetadata = true
-					}
-				}
+func (vs *volumeSnapshotter) findVolumeSnapshotMetadata(snapshot *crdv1.VolumeSnapshot) *map[string]string {
+	var tags *map[string]string = nil
+	if snapshot.Metadata.Name != "" && snapshot.Metadata.Namespace != "" {
+		if snapshot.Metadata.Labels != nil {
+			timestamp, ok := snapshot.Metadata.Labels["Timestamp"]
+			if ok {
+				tags := make(map[string]string)
+				tags[CloudSnapshotCreatedForVolumeSnapshotNamespaceTag] = snapshot.Metadata.Namespace
+				tags[CloudSnapshotCreatedForVolumeSnapshotNameTag] = snapshot.Metadata.Name
+				tags[CloudSnapshotCreatedForVolumeSnapshotTimestampTag] = timestamp
+				glog.Infof("findVolumeSnapshotMetadata: returning tags [%#v]", tags)
 			}
+		}
+	}
+	return tags
+}
 
-			// If metadata not found, create it in VolumeSnapshot now
-			if !bFoundVolumeSnapshotMetadata {
-				tags, err = vs.UpdateVolumeSnapshot(snapshotName, nil, bFoundVolumeSnapshotMetadata)
-			}
+func (vs *volumeSnapshotter) getPlugin(snapshotName string, snapshot *crdv1.VolumeSnapshot) (*volume.VolumePlugin, error) {
+	pv, err := vs.getPVFromVolumeSnapshot(snapshotName, snapshot)
```

Comment:
This is not safe: the PV must not be consulted to get a plugin name (or only when creating the snapshot) since PV is not bound to the snapshot in any way: it may (and will -- it's one of the snapshot use-cases!) be deleted. We should really remove the PV reference from the VolumeSnapshotDataSpec...

Reply:
There is a TODO item to cache PV volume information into VolumeSnapshot meta data to avoid querying api server.  Does that address your concern?